crates/walrus-indexer/src/storage.rs

```diff
@@ -0,0 +1,471 @@
+    /// Optimized parsing: first 32 bytes = blob_manager, last 8 bytes = sequence,
+    /// middle = /<identifier>/.
+    fn from_key(bytes: &[u8]) -> Result<Self> {
+        // Minimum size: 32 (blob_manager) + 1 (/) + 0 (empty identifier) + 1 (/) + 8 (sequence).
+        if bytes.len() < 42 {
+            return Err(anyhow::anyhow!("Key too short: minimum 42 bytes required"));
+        }
+
+        // Extract blob_manager (first 32 bytes).
+        let blob_manager = ObjectID::from_bytes(&bytes[0..32])?;
+
+        // Check for first separator.
+        if bytes[32] != b'/' {
+            return Err(anyhow::anyhow!(
+                "Missing first '/' separator after blob_manager"
+            ));
+        }
+
+        // Extract sequence number (last 8 bytes).
+        let seq_bytes: [u8; Self::SEQUENCE_LEN] =
+            bytes[bytes.len() - Self::SEQUENCE_LEN..].try_into()?;
+        let sequence_number = u64::from_be_bytes(seq_bytes);
+
+        // Check for second separator before sequence.
+        if bytes[bytes.len() - Self::SEQUENCE_LEN - 1] != b'/' {
+            return Err(anyhow::anyhow!(
+                "Missing second '/' separator before sequence"
+            ));
+        }
+
+        // Extract identifier (everything between the two separators).
+        // From position 33 to (length - 9).
+        let identifier =
+            String::from_utf8(bytes[33..bytes.len() - Self::SEQUENCE_LEN - 1].to_vec())?;
+
+        Ok(Self {
+            blob_manager,
+            identifier,
+            sequence_number,
+        })
+    }
+}
+
+impl std::fmt::Display for PrimaryIndexKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}/{}/{}",
+            self.blob_manager, self.identifier, self.sequence_number
+        )
+    }
+}
+
+/// Key structure for object index.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct ObjectIndexKey {
+    object_id: ObjectID,
+}
+
+impl ObjectIndexKey {
+    fn new(object_id: ObjectID) -> Self {
+        Self { object_id }
+    }
+
+    /// Convert to optimized byte representation for RocksDB key.
+    /// Format: [object_id_bytes(32 bytes)].
+    fn to_key(&self) -> Vec<u8> {
+        // Simply use the raw ObjectID bytes (32 bytes).
+        self.object_id.as_ref().to_vec()
+    }
+
+    /// Parse from byte representation.
+    #[allow(unused)]
+    fn from_key(bytes: &[u8]) -> Result<Self> {
+        if bytes.len() != 32 {
+            return Err(anyhow::anyhow!(
+                "Invalid object_id length: expected 32 bytes"
+            ));
+        }
+        Ok(Self {
+            object_id: ObjectID::from_bytes(bytes)?,
+        })
+    }
+}
+
+impl std::fmt::Display for ObjectIndexKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.object_id)
+    }
+}
+
+/// Key structure for quilt patch index.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct QuiltPatchIndexKey {
```

Comment:
Question: what's the relationship between QuiltPatchIndexKey and PrimaryIndexKey?